Storage/SQL/SQLiteHistoryRecommendations.swift

```diff
@@ -0,0 +1,97 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+import Shared
+import XCGLogger
+import Deferred
+
+/// Small enum to help strength parameter requirements for potential bookmark highlights query below.
+private enum BookmarkSource {
+    case Mirror
+    case Local
+
+    var modifiedColumn: String  {
+        switch self {
+        case Mirror: return "server_modified"
+        case Local: return "local_modified"
+        }
+    }
+
+    var tableName: String {
+        switch self {
+        case Mirror: return TableBookmarksMirror
+        case Local: return TableBookmarksLocal
+        }
+    }
+}
+
+extension SQLiteHistory: HistoryRecommendations {
+    public func getHighlights() -> Deferred<Maybe<Cursor<Site>>> {
+        let limit = 19
+        let microsecondsPerMinute: Timestamp = 60_000_000 // 1000 * 1000 * 60
+        let thirtyMinutesAgo = NSDate.nowMicroseconds() - 30 * microsecondsPerMinute
+        let threeDaysAgo = NSDate.nowMicroseconds() - (60 * microsecondsPerMinute) * 24 * 3
+        let bookmarkLimit = 1
+        let historyLimit = limit - bookmarkLimit
+
+        let recentHistory = [
+            "SELECT *",
+            "FROM (",
+            "   SELECT \(TableHistory).id AS historyID, url, title, guid, sum(1) AS visitCount, max(\(TableVisits).date) AS visitDate",
+            "   FROM \(TableHistory)",
+            "   LEFT JOIN \(TableVisits) ON \(TableVisits).siteID = \(TableHistory).id",
+            "   WHERE title NOT NULL AND title != '' AND is_deleted = 0",
+            "   GROUP BY url",
+            "   ORDER BY visitDate DESC",
+            "   LIMIT \(historyLimit)",
+            ")",
+            "WHERE visitCount <= 3 AND visitDate < \(thirtyMinutesAgo)",
```

Comment:
Two problems with this.

Firstly, I think your comparison is backwards — you're asking for timestamps that are numerically less than a moment in time, so this will be *non-recent* history (smaller timestamp).

Secondly, this is a hopelessly inefficient way of phrasing this. Presumably there aren't many sites visited in the last thirty minutes, so make that your most specific clause. Something like:

```
SELECT DISTINCT siteID FROM \(TableVisits) WHERE date > \(thirtyMinutesAgo)
```

is a nice subquery to give you all recently visited sites.

Because you always know the most recent date will be in the candidate rows, you can get the max date for free:

```
SELECT siteID, max(date) AS latest FROM \(TableVisits) WHERE date > \(thirtyMinutesAgo) GROUP BY siteID;
```

You'd need to join back against the visits table to find the count. You use `sum` to do this, but I'd use a subquery in the projection:

```
SELECT s AS siteID, latest, (SELECT COUNT(1) FROM \(TableVisits) WHERE siteID = s) AS visitCount
FROM
(SELECT siteID AS s, max(date) AS latest
 FROM \(TableVisits) WHERE date > \(thirtyMinutesAgo)
 GROUP BY siteID
 LIMIT \(historyLimit))
WHERE visitCount <= 3;
```